### DISCUSSION TITLE
--log-prefix=pad

### DISCUSSION THREAD

**Discussion by gajus** [vercel/turborepo#7484] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

Pad logs to the length of the longest workspace+task name.

Consider these logs:

```
contra-temporal:dev: cache bypass, force executing 3d7c2284345252a4
contra-temporal:dev:
contra-temporal:dev: > contra-temporal@ dev /Users/gajus/Developer/contra/gaia-og1/apps/contra-temporal
contra-temporal:dev: > docker compose --profile temporal up
contra-temporal:dev:
contra-redis:dev:
contra-redis:dev: > contra-redis@ dev /Users/gajus/Developer/contra/gaia-og1/apps/contra-redis
contra-redis:dev: > docker run --name contra-redis --rm -p 6382:6379 redis
contra-redis:dev:
contra-database:dev:
contra-database:dev: > contra-database@ dev /Users/gajus/Developer/contra/gaia-og1/apps/contra-database
contra-database:dev: > docker build -t contra-database -f Dockerfile --build-arg DIFF=false ../.. && docker run -p 5435:5432 --name contra-database --rm contra-database
contra-database:dev:
contra-redis:dev: docker: Error response from daemon: Conflict. The container name "/contra-redis" is already in use by container "ad4d87a63c04d75c6c0f117275a261153c810cbbb2b13aabda8c0aaaf5868d74". You have to remove (or rename) that container to be able to reuse that name.
contra-redis:dev: See 'docker run --help'.
contra-redis:dev:  ELIFECYCLE  Command failed with exit code 125.
contra-redis:dev:  WARN   Local package.json exists, but node_modules missing, did you mean to install?
contra-redis:dev: ERROR: command finished with error: command (/Users/gajus/Developer/contra/gaia-og1/apps/contra-redis) pnpm run dev exited (1)
```

If `contra-temporal:dev` is the longest name in the repo, then with the proposed option, these logs would be adjusted to:

```
contra-temporal:dev: cache bypass, force executing 3d7c2284345252a4
contra-temporal:dev:
contra-temporal:dev: > contra-temporal@ dev /Users/gajus/Developer/contra/gaia-og1/apps/contra-temporal
contra-temporal:dev: > docker compose --profile temporal up
contra-temporal:dev:
contra-redis:dev   :
contra-redis:dev   : > contra-redis@ dev /Users/gajus/Developer/contra/gaia-og1/apps/contra-redis
contra-redis:dev   : > docker run --name contra-redis --rm -p 6382:6379 redis
contra-redis:dev   :
contra-database:dev:
contra-database:dev: > contra-database@ dev /Users/gajus/Developer/contra/gaia-og1/apps/contra-database
contra-database:dev: > docker build -t contra-database -f Dockerfile --build-arg DIFF=false ../.. && docker run -p 5435:5432 --name contra-database --rm contra-database
contra-database:dev:
contra-redis:dev   : docker: Error response from daemon: Conflict. The container name "/contra-redis" is already in use by container "ad4d87a63c04d75c6c0f117275a261153c810cbbb2b13aabda8c0aaaf5868d74". You have to remove (or rename) that container to be able to reuse that name.
contra-redis:dev   : See 'docker run --help'.
contra-redis:dev   :  ELIFECYCLE  Command failed with exit code 125.
contra-redis:dev   :  WARN   Local package.json exists, but node_modules missing, did you mean to install?
contra-redis:dev   : ERROR: command finished with error: command (/Users/gajus/Developer/contra/gaia-og1/apps/contra-redis) pnpm run dev exited (1)
```

It just makes reading logs easier.

### Describe the solution you'd like

N/A

### Describe alternatives you've considered

N/A

**Comment by mehulkar** [comment]
`--log-prefix=pad` would probably not be enough to express this, since we would need it to work for other prefix formats that may be added. E.g. we already have [an implicit log-prefix for "single package repos")][1]. If we made that explicit, we would need it to also work with padding.

[1]: https://github.com/vercel/turbo/blob/451eb14eab3d31bbe74204ff0fc4ab442bcfee51/cli/internal/nodes/packagetask.go#L25-L30

I had considered the possibility of expanding `--log-prefix` to allow a microsyntax (like `git log --pretty=format:...`) with placeholders. In that world, we would have placeholders to represent the taskID, taskName, workspace name, etc. That could maybe include a placeholder for padding. For this request, you could be expressed as:

```sh
# %i represents taskID
# %p represents "pad", so appearing at the end would be "right pad"
turbo run --log-prefix='format:%i%p

# %p at the beginning would be left-pad
turbo run --log-prefix='format:%p%i
```

What do you think of that? I'm not sure how/if we'd allow `%p` to be at both ends, or somewhere in the middle

**Comment by gajus** [comment]
That would be cool!

Alternatively... and this could be a stretch, but still want to suggest it – you could support what [@roarr/cli](https://github.com/gajus/roarr-cli#roarr-configuration-file) does and allow a configuration file with log formatting rules. I'd prefer this because it would allow to perform other useful features, like filtering, throttling, resolving paths in logs, etc. in user-land.

**Comment by mehulkar** [comment]
Structured logging all around is a great idea. We are currently planning for an upcoming release to take a look at logging holistically. We can take a look at this if/when that sprint happens.

**Comment by gajus** [comment]
Structured logging is great, but in this case, I simply meant a config file that defines a callback that gets unstructured log message that I can post-process myself.
